dln/operator.py

```diff
@@ -13,74 +12,124 @@
+    return output
+
+
+class LLM(ABC):
+
+    def __init__(self, model_name: str, **generation_options):
+        self.generation_options = generation_options
+        self.engine = model_name
+
+    def __call__(self, inputs: Union[List[str], str], **kwargs) -> List[str]:
+        return self.generate(inputs, **kwargs)
+
+    @abstractmethod
+    def generate(self, inputs: Union[List[str], str], **kwargs) -> List[str]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def encode(self, string: str) -> List[int]:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def has_log_probs(self) -> bool:
+        raise NotImplementedError
+    
+    def compute_cost(self, inputs: List[str]) -> float:
+        return np.sum(list([len(self.encode(input)) for input in inputs]))
+
+
+class GPT(LLM):
+
+    CHAT_COMPLETION_MODELS = [
+        "gpt-35-turbo",  # azure
+        "gpt-3.5-turbo",
+        "gpt-4",
+        "gpt-4-32k",
+        "gpt-4-0613",
+    ]
 
-class GPT:
-    AVAILABLE_MODELS = [
+    COMPLETION_MODELS = [
         "text-davinci-003",
         "text-davinci-002",
         "code-davinci-002",
         "text-curie-001",
         "text-babbage-001",
         "text-ada-001",
-        "gpt-3.5-turbo",
-        "gpt-35-turbo",
-        "gpt-4",
-        "gpt-4-32k",
-        "gpt-4-0613",
-        "any",
     ]
 
+    AVAILABLE_MODELS = CHAT_COMPLETION_MODELS + COMPLETION_MODELS
+
     def __init__(self, model_name="text-davinci-003", **generation_options):
         if model_name not in self.AVAILABLE_MODELS:
             raise ValueError(
-                f"model_name should be one of: {','.join(self.AVAILABLE_MODELS)}"
+                f"GPT model_name should be one of: {','.join(self.AVAILABLE_MODELS)}"
             )
-
-        self.generation_options = generation_options
-        self.engine = model_name
-
-        if self.engine == "any":
-            openai.api_base = "http://0.0.0.0:8081"
-            openai.api_key = "any"
-            openai.api_type = "openai"
-            self.encoder = tiktoken.encoding_for_model("text-davinci-003")
-        else:
-            self.encoder = tiktoken.encoding_for_model(self.engine)
+        super().__init__(model_name, **generation_options)
+        engine_for_encoder = self.engine
+        if engine_for_encoder == "gpt-35-turbo":
+            engine_for_encoder = "gpt-3.5-turbo"
+        self.encoder = instantiate_tokenizer(engine_for_encoder)
         openai.api_version = os.environ.get('OPENAI_API_VERSION')
 
-    def encode(self, string):
+    def encode(self, string: str) -> List[int]:
         return self.encoder.encode(string)
 
     @property
-    def has_log_probs(self):
-        return not (
-            "gpt-3.5" in self.engine or
-            "gpt-4" in self.engine or
-            "gpt-35" in self.engine or
-            "gpt-4-0613" in self.engine
-        )
+    def has_log_probs(self) -> bool:
+        return self.engine in self.COMPLETION_MODELS
```

Comment:
In the future, non completion models could also offer logprobs. Instead, we could have another variable to keep track of which model supports logprobs?